pex_reactor.go

```diff
@@ -61,31 +76,37 @@ func (pexR *PEXReactor) GetChannels() []*ChannelDescriptor {
 	}
 }
 
-// Implements Reactor
-func (pexR *PEXReactor) AddPeer(peer *Peer) {
-	// Add the peer to the address book
-	netAddr := NewNetAddressString(peer.ListenAddr)
-	if peer.IsOutbound() {
-		if pexR.book.NeedMoreAddrs() {
-			pexR.RequestPEX(peer)
+// AddPeer implements Reactor by adding peer to the address book (if inbound)
+// or by requesting more addresses (if outbound).
+func (r *PEXReactor) AddPeer(p *Peer) {
+	if p.IsOutbound() { // For outbound peers, the address is already in the books
+		if r.book.NeedMoreAddrs() {
+			r.RequestPEX(p)
 		}
-	} else {
-		// For inbound connections, the peer is its own source
-		// (For outbound peers, the address is already in the books)
-		pexR.book.AddAddress(netAddr, netAddr)
+	} else { // For inbound connections, the peer is its own source
+		addr := NewNetAddressString(p.ListenAddr)
+		r.book.AddAddress(addr, addr)
 	}
 }
 
-// Implements Reactor
-func (pexR *PEXReactor) RemovePeer(peer *Peer, reason interface{}) {
-	// TODO
+// RemovePeer implements Reactor by removing peer from the address book.
+func (r *PEXReactor) RemovePeer(p *Peer, reason interface{}) {
+	addr := NewNetAddressString(p.ListenAddr)
+	// addr will be ejected from the book
+	r.book.MarkBad(addr)
```

Comment:
But if the peer actually went offline, wont all the other peers remove him too? Granted the PEX should enable everyone to find him when he next connects.